README.md
# Player JS
## _The Most Customized Player For Web Devlopement_

![Image](https://i.imgur.com/ZoFCFrg.png)

PlayerJS is video player for web created with javascript. This Package was created to make you able to use it from your Dart code as a Widget for Your Web App.
## Installation


- You Must To Create Your Own Player in [PlayerJS](https://playerjs.com/)
- Then Follow The Instructions To Make It
- After You Complete Make Sure to Download The Player
- Then Put It In Web Directory In Your Flutter Project
- After That Import It In index.html using script Tag
- Now You Are  Ready To Use The Package

## Features

- Supported Live Streaming
- Very Fast
- High Quality Codecs (H.264 + AAC)
- Chromecast
- Google Analytics
- Support Advertising
- Separate Iframe

## Supported Formats

You Need To Enable All Features First When You Build The Player .

| Name | Extensions |
| ------ | ------ |
| MPEG-4 | .MP4|
| Matroska | .MKV |
| Audio Video Interleave | .AVI |
| WebM | .WEBM |
| HLS | .M3U8 |
| MPEG-DASH | .MPD |

## Example

```dart
Player(
  videoUrl:
  'http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4',
  subtitles: {
    'ar': 'https://example/arabic.srt',
    'en': 'https://example/english.srt'
  },
  height: 400,
  width: 700,
)
```

## BSD 3-Clause License


```
BSD 3-Clause License

Copyright (c) 2023 Djeddi Yacine
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```

CHANGELOG.md
## 1.0.0

* First Release.

## 1.0.1

* Fix Some Code

## 1.0.2

* Remove license keyword from pubspec

## 1.0.3

* change the license

## 1.0.4

* remove universal_html

test/player_js_test.dart
import 'package:player_js/player_js.dart';

void main() {
  const Iframe(
    url:
        'http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4',
    width: 600,
    height: 300,
    pathToPlayer: '',
    subtitles: {},
    poster: '',
    title: '',
    viewType: 'iframe',
  );
  const Player(
    height: 400,
    width: 700,
    videoUrl:
        'http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4',
    subtitles: {},
  );
}


lib/src/player.dart
part of player_js;

///Render Video with Custom PlayerJS
class Player extends StatefulWidget {
  ///  Add the Height of the Video
  ///  like : 400
  final double height;

  ///  Add the Width of the Video
  ///  like : 700
  final double width;

  ///  Add the Direct URL of the video
  ///  like :  'https://example/video/video.mp4'
  final String videoUrl;

  /// add subtitles as a map like :
  ///   {'AR':'https://example/video/arabic.srt',
  ///   'EN':'https://example/video/english.vtt'}
  ///  remove all space in the ''
  final Map<String, String> subtitles;
  const Player({
    Key? key,
    required this.height,
    required this.width,
    required this.videoUrl,
    required this.subtitles,
  }) : super(key: key);

  @override
  State<Player> createState() => _PlayerState();
}

class _PlayerState extends State<Player> {
  late final String viewType = 'video-player-${widget.hashCode}';
  late final String playerId = 'player-${widget.hashCode}';
  late final String div = 'div-${widget.hashCode}';
  late String subtitleString = '';
  @override
  void initState() {
    super.initState();
    PlayerConfig.init();
    PlayerConfig.register(
      viewType: viewType,
      playerId: playerId,
      div: div,
      subtitleString: subtitleString,
      hashCode: widget.hashCode,
    );
    widget.subtitles.forEach((lang, url) {
      subtitleString += '[$lang]$url,';
    });
  }

  @override
  Widget build(BuildContext context) {
    if (kIsWeb) {
      // Create a platform view for the video player
      return Container(
        color: Colors.transparent,
        height: widget.height,
        width: widget.width,
        child: HtmlElementView(
          viewType: viewType,
          onPlatformViewCreated: (int id) {
            js.context.callMethod(
                'setupPlayer', [widget.videoUrl, playerId, subtitleString]);
          },
        ),
      );
    } else {
      return Container(
        color: Theme.of(context).primaryColor,
        width: widget.width,
        height: widget.height,
        child: const Center(child: Text('Unsupported Platform')),
      );
    }
  }
}


lib/src/iframe_config.dart
part of player_js;

class IframeConfig {
  static void init() {
    if (!kIsWeb) {
      throw UnsupportedError('Iframe only supports web');
    }
  }

  static void registerView({
    required String viewType,
    required String pathToPlayer,
    required String url,
    required double width,
    required double height,
    required Map<String, String> subtitles,
    required String poster,
    required String title,
  }) {
    {
      String buildUrl(
        String pathToPlayer,
        String url,
        Map<String, String>? subtitles,
        String? poster,
        String? title,
      ) {
        final subtitleList = subtitles?.entries
            .map((entry) => '[${entry.key}]${entry.value}')
            .join(',');
        return '$pathToPlayer?file=$url&poster=${poster ?? ''}&title=${title ?? ''}&subtitle=${subtitleList ?? ''}';
      }

      // ignore: undefined_prefixed_name
      ui.platformViewRegistry.registerViewFactory(
        viewType,
        (int viewId) => IFrameElement()
          ..height = '100%'
          ..width = '100%'
          ..style.height = '100%'
          ..style.width = '100%'
          ..src = buildUrl(pathToPlayer, url, subtitles, poster, title),
      );
    }
  }
}


lib/src/player_config.dart
part of player_js;

class PlayerConfig {
  static void init() {
    if (!kIsWeb) {
      throw UnsupportedError('PlayerJs only supports web');
    }
  }

  static void register({
    required Object hashCode,
    required String viewType,
    required String playerId,
    required String div,
    required String subtitleString,
  }) {
    {
      // ignore: undefined_prefixed_name
      ui.platformViewRegistry.registerViewFactory(viewType, (int viewId) {
        final big = DivElement()
          ..id = div
          ..style.height = '100%'
          ..style.width = '100%';

        final playerElement = DivElement()
          ..id = playerId
          ..style.height = '100%'
          ..style.width = '100%';

        final script = ScriptElement()
          ..type = 'text/javascript'
          ..text = '''
          function setupPlayer(videoUrl, playerId, subtitle) {
            var player = new Playerjs({
              "id": playerId,
              "file": videoUrl,
              "subtitle": subtitle,
            });
          }
        ''';

        big.append(playerElement).append(script);
        return big;
      });
    }
  }
}


lib/src/iframe.dart
part of player_js;

///Render Iframe with Custom PlayerJS
class Iframe extends StatefulWidget {
  final String viewType;

  /// Add the path of the player.html.
  /// make sure that the same folder
  /// contain both the player.html
  /// and playerjs.js
  /// put the folder in the web
  /// folder in the root of your project
  final String pathToPlayer;

  ///  Add the Direct URL of the video
  ///  like :  'https://example/video/video.mp4'
  final String url;

  ///  Add the Width of the iframe
  ///  like : 700
  final double width;

  ///  Add the Height of the iframe
  ///  like : 400
  final double height;

  /// add subtitles as a map like :
  ///   {'AR':'https://example/video/arabic.srt',
  ///   'EN':'https://example/video/english.vtt'}
  ///  remove all space in the ''
  final Map<String, String> subtitles;

  /// add Poster like  :
  /// 'https://example/video/poster.png'
  final String poster;

  /// add Title like : 'Avengers'
  final String title;

  const Iframe({
    Key? key,
    required this.pathToPlayer,
    required this.url,
    required this.width,
    required this.height,
    required this.subtitles,
    required this.poster,
    required this.title,
    required this.viewType,
  }) : super(key: key);

  @override
  State<Iframe> createState() => _IframeState();
}

class _IframeState extends State<Iframe> {
  @override
  void initState() {
    super.initState();
    IframeConfig.init(); // initialize the plugin
    IframeConfig.registerView(
      viewType: widget.viewType,
      pathToPlayer: widget.pathToPlayer,
      url: widget.url,
      width: widget.width,
      height: widget.height,
      subtitles: widget.subtitles,
      poster: widget.poster,
      title: widget.title,
    );
  }

  @override
  Widget build(BuildContext context) {
    if (kIsWeb) {
      return Container(
        width: widget.width,
        height: widget.height,
        color: Colors.transparent,
        child: HtmlElementView(viewType: widget.viewType),
      );
    } else {
      return Container(
        color: Theme.of(context).primaryColor,
        width: widget.width,
        height: widget.height,
        child: const Center(child: Text('Unsupported Platform')),
      );
    }
  }
}
